packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { TypedEventEmitter, assert } from "@fluidframework/common-utils";
+import { IConnectionDetails, IDeltaHandlerStrategy, IDeltaQueue, IDeltaQueueEvents } from "@fluidframework/container-definitions";
+import { IDocumentService } from "@fluidframework/driver-definitions";
+import { IDocumentMessage, ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IConnectionManager, IConnectionManagerFactoryArgs } from "./contracts";
+import { DeltaManager } from "./deltaManager";
+
+const GroupedBatchOpType = "groupedBatch";
+
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
+    private readonly _trackedListeners: Map<string, (...args: any[]) => void> = new Map();
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T, storeValue?: boolean) => T,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => { this.emit(event, this._virtualizeValue(task, true /* storeValue */)); }
```

Comment:
General note: it's easy to reason about what needs to happen when direction of communication is clear. I.e. runtime pushes ops into the queue.
It's hard to impossible when
- direction is unknown (for example, "push" - is this result of runtime pushing (indirectly) into outbound queue, or it's connection manager pushing received ops into inbound queue
- methods like toArray, where we do not know anything about consumer (and direction of the flow).

So I have a feeling that doing transformations here is wrong. It's very hard to ensure that we covered all the ways someone can access content (i.e. why do we convert here on "op" but not on peek()?)

Plus "push" for inbound & outbound queues needs different type of transformation, right?

Instead, I think we should push responsibility to do conversions to next layer (DM wrapper) when content is pushed into the queue, or when content is consumed from the queue.

Reply:
Oh, I realized that one point is not valid. How about changing _virtualizeValue work to "transform"? As it can go both directions.